pkg/controller/nodeimage/nodeimage_controller.go

```diff
@@ -0,0 +1,468 @@
+
+// ReconcileNodeImage reconciles a NodeImage object
+type ReconcileNodeImage struct {
+	client.Client
+	scheme        *runtime.Scheme
+	clock         clock.Clock
+	eventRecorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=nodeimages,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=nodeimages/status,verbs=get;update;patch
+
+// Reconcile reads that state of the cluster for a NodeImage object and makes changes based on the state read
+// and what is in the NodeImage.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write nodes
+func (r *ReconcileNodeImage) Reconcile(request reconcile.Request) (res reconcile.Result, err error) {
+	start := time.Now()
+	klog.V(5).Infof("Starting to process NodeImage %v", request.Name)
+	var requeueMsg string
+	defer func() {
+		if err != nil {
+			klog.Warningf("Failed to process NodeImage %v err %v, elapsedTime %v", request.Name, time.Since(start), err)
+		} else if res.RequeueAfter > 0 {
+			klog.Infof("Finish to process NodeImage %v, elapsedTime %v, RetryAfter %v: %v", request.Name, time.Since(start), res.RequeueAfter, requeueMsg)
+		} else {
+			klog.Infof("Finish to process NodeImage %v, elapsedTime %v", request.Name, time.Since(start))
+		}
+	}()
+
+	// Fetch the Node
+	node := &v1.Node{}
+	err = r.Get(context.TODO(), request.NamespacedName, node)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to get node %s: %v", request.NamespacedName.Name, err)
+		}
+		node = nil
+	}
+
+	// Fetch the NodeImage
+	nodeImage := &appsv1alpha1.NodeImage{}
+	err = r.Get(context.TODO(), request.NamespacedName, nodeImage)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to get nodeimage %s: %v", request.NamespacedName.Name, err)
+		}
+		nodeImage = nil
+	}
+
+	// If Node not exists or has been deleted
+	if node == nil || node.DeletionTimestamp != nil {
+		// All been deleted
+		if nodeImage == nil || nodeImage.DeletionTimestamp != nil {
+			return reconcile.Result{}, nil
+		}
+
+		if _, ok := nodeImage.Labels[fakeLabelKey]; !ok {
+			// Delete nodeImage
+			if err = r.Delete(context.TODO(), nodeImage); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to delete nodeimage %v, err: %v", nodeImage.Name, err)
+			}
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// If Node exists, we should create a NodeImage
+	if nodeImage == nil {
+		nodeImage = &appsv1alpha1.NodeImage{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: request.Name,
+			},
+			Spec: appsv1alpha1.NodeImageSpec{},
+		}
+		if err = r.Create(context.TODO(), nodeImage); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to create nodeimage %v, err: %v", nodeImage.Name, err)
+		}
+		klog.Infof("Successfully create nodeimage %s", nodeImage.Name)
```

Comment:
just return  anyway since no need to update image status.

Reply:
There is no need to update spec here. nodeimage-controller only creates the nodeimage object, and imagepulljob-controller will add images needed to pull into it.